src/object/cli_obj.c

```diff
@@ -994,6 +995,22 @@ obj_shard_tgts_query(struct dc_object *obj, uint32_t map_ver, uint32_t shard,
 		D_GOTO(out, rc);
 	}
 
+	shard_tgt->st_ec_tgt = -1;
+	if (bitmap != NIL_BITMAP) {
+		uint32_t tgt_idx;
+		uint32_t grp_idx;
+
+		grp_idx = shard / obj_get_grp_size(obj);
+		tgt_idx = obj_shard->do_id.id_shard -
```

Comment:
shard is the real offset of the shard within the layout array (dc_object->cob_shards).  id_shard is the shard id of the object.   They are different only if there are extending shards during online reintegration, since we need write data to both and new shards,  which share the same id_shard,  but different offset in layout array.

tgt_idx is the id_shard offset within one EC group.

Reply:
I mean why not "tgt_idx = shard - grp_idx * daos_oclass_grp_size(obj_get_oca(obj_auxi->obj));" ?

Reply:
tgt_idx is the shard_id within one EC group. 

 If the shard is being extended (online reintegration), there will be multiple shards point to the same shard id.    Let's say the object EC_2P1G1  is being reintegrated, so the shard layout would be

shard_idx,        shard_id,      tgt_id,
0,                        0,                     1,
1,                         1,                     2,
2,                        2,                    3,
3,                        1,                     4,               <--- this shard(shard 1) is being extended,  so shard_idx(offset) is 3,  and shard_id is 1, meaning the shard_id 1 is migrated from 2 to 4.

>tgt_idx = shard - grp_idx * daos_oclass_grp_size(obj_get_oca(obj_auxi->obj));

So in this case, shard(offset) is 3,  id_shard is 1,   but tgt_idx is supposed to be the shard_id within that bitmap (all about implementation details).  So we need use shard_id, instead of shard.

Reply:
> So in this case, shard(offset) is 3, id_shard is 1, but tgt_idx is supposed to be the shard_id within that bitmap (all about implementation details). So we need use shard_id, instead of shard.

Assume all shards are updated, then what the bitmap looks like? "111" or "1111"?

Reply:
111